src/ServiceControl.Monitoring/Diagrams/MonitoringData.cs

```diff
@@ -0,0 +1,55 @@
+﻿namespace ServiceControl.Monitoring.Raw
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Threading;
+
+    public class MonitoringData
+    {
+        public MonitoringData(int historySize)
+        {
+            this.historySize = historySize;
+
+            Endpoints = new ConcurrentDictionary<string, EndpointData>();
+        }
+
+        public EndpointData Get(string endpointName)
+        {
+            return Endpoints.GetOrAdd(endpointName, new EndpointData(historySize));
+        }
+
+        public ConcurrentDictionary<string, EndpointData> Endpoints;
+        readonly int historySize;
+    }
+
+    public class EndpointData
+    {
+        public EndpointData(int size)
+        {
+            Timestamps = new DateTime[size];
+            CriticalTime = new float?[size];
+            ProcessingTime = new float?[size];
+
+            this.size = size;
+            head = 0;
+        }
+
+        public void Record(DateTime timestamp, float? criticalTime, float? processingTime)
+        {
+            Interlocked.Increment(ref head);
+
+            var index = (head - 1) % size;
```

Comment:
Use the value returned from `Interlocked.Increment` rather than rereading the field (which depending on various conditions might return a different value)